benchmark/Benchmark.h

```diff
@@ -0,0 +1,172 @@
+// Copyright 2022, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (November of 2022, schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <vector>
+#include <functional>
+#include <string>
+
+#include "util/Timer.h"
+#include <util/HashMap.h>
+
+/*
+ * Used for measuring the time needed for the execution of a function. It also
+ * saves the measured time togehter with a description of the function
+ * measured.
+ */
+class BenchmarkRecords {
+
+  public:
+
+    // Describes a measured function.
+    struct RecordEntry {
+      std::string descriptor; // Needed, because without it, nobody could tell,
+                              // which time belongs to which benchmark.
+      float measuredTime; // The measured time in seconds.
+    };
+
+    // Describes a group of measured functions.
+    struct RecordGroup {
+      std::string descriptor; // Needed for identifying groups.
+      std::vector<RecordEntry> entries;
+    };
+    
+    // Describes a table of measured functions.
+    struct RecordTable {
+      // For identification.
+      std::string descriptor;
+      // The names of the columns and rows.
+      std::vector<std::string> rowNames;
+      std::vector<std::string> columnNames;
+      // The actual content of the fields. As in the seconds one function
+      // execution takes. Access is [row, column].
+      std::vector<std::vector<float>> entries;
+    };
+
+  private:
+
+    // A vector of all single functions measured.
+    std::vector<RecordEntry> _singleMeasurments;
+
+    // A hash map of all the created RecordGroups. For faster access.
+    // The key for a RecordGroup is it's descriptor.
+    ad_utility::HashMap<std::string, RecordGroup> _recordGroups;
+  
+    // A hash map of all the created RecordTables. For faster access.
+    // The key for a RecordTable is it's descriptor.
+    ad_utility::HashMap<std::string, RecordTable> _recordTables;
+    
+    /*
+     * @brief Return execution time of function in seconds.
+     */
+    float measureTimeOfFunction(const std::function<void()>& functionToMeasure) const;
+
```

Comment:
I would suggest making all the functions that measure the time (this one, the `addSingleMeasurement`, `addToExistingGroup` and `addToExistingTable`) templates that are templated on the type of `functionToMeasure`. You would then have to define them directly in the header. This reduces the measurement overhead of the Benchmarking interface (`std::function` has quite some indirections).

You could add a concept/requires clause to check that the function is still callable with no arguments and that it returns void, but this can be done in a second step, I can help you with that.

Reply:
Done, see [901fc2e](https://github.com/ad-freiburg/qlever/pull/860/commits/901fc2e9fd858394e35bd68738890df3fcecbe42).

Reply:
I implemented the requirements, after reading up on the topic for a different task. See [44ef1ec](https://github.com/ad-freiburg/qlever/pull/860/commits/44ef1ecb8d6af0f3907068faa81c0b6e5542891c).